Pegasus/sources/Geometry.cpp

```diff
@@ -6,182 +6,257 @@
+        = intersection::CalculatePenetration<Sphere, Plane>;
+    m_calculatePenetrationFunctors[std::make_pair(SimpleShape::Type::SPHERE, SimpleShape::Type::SPHERE)]
+        = intersection::CalculatePenetration<Sphere, Sphere>;
+    m_calculatePenetrationFunctors[std::make_pair(SimpleShape::Type::SPHERE, SimpleShape::Type::BOX)]
+        = intersection::CalculatePenetration<Sphere, Box>;
+    m_calculatePenetrationFunctors[std::make_pair(SimpleShape::Type::BOX, SimpleShape::Type::PLANE)]
+        = intersection::CalculatePenetration<Box, Plane>;
+    m_calculatePenetrationFunctors[std::make_pair(SimpleShape::Type::BOX, SimpleShape::Type::SPHERE)]
+        = intersection::CalculatePenetration<Box, Sphere>;
+    m_calculatePenetrationFunctors[std::make_pair(SimpleShape::Type::BOX, SimpleShape::Type::BOX)]
+        = intersection::CalculatePenetration<Box, Box>;
 }
 
-pegasus::geometry::Cylinder::Cylinder(
-    glm::dvec3 const& centerOfMass, glm::dvec3 const& halfHeight, double r
-)
-    : Capsule(centerOfMass, halfHeight, r)
+void pegasus::geometry::SimpleShapeIntersectionDetector::Initialize(SimpleShape const* a, SimpleShape const* b)
 {
+    m_initializeFunctors[std::make_pair(a->type, b->type)](
+        a, b, m_intersectionCaches[std::make_pair(a->type, b->type)].get());
 }
 
-pegasus::geometry::Box::Box(
-    glm::dvec3 const& centerOfMass, glm::dvec3 const& a, glm::dvec3 const& b, glm::dvec3 const& c
-)
-    : SimpleShape(centerOfMass, SimpleShapeType::BOX)
-    , m_aAxis(a)
-    , m_bAxis(b)
-    , m_cAxis(c)
+bool pegasus::geometry::SimpleShapeIntersectionDetector::CalculateIntersection(SimpleShape const* a, SimpleShape const* b)
 {
+    return m_calculateIntersectionFunctors[std::make_pair(a->type, b->type)](
+        a, b, m_intersectionCaches[std::make_pair(a->type, b->type)].get());
 }
 
-void pegasus::geometry::Box::SetAxes(glm::dvec3 const& a, glm::dvec3 const& b, glm::dvec3 const& c)
+glm::dvec3 pegasus::geometry::SimpleShapeIntersectionDetector::CalculateContactNormal(SimpleShape const* a, SimpleShape const* b)
 {
-    m_aAxis = a;
-    m_bAxis = b;
-    m_cAxis = c;
+    return m_calculateContactNormalFunctors[std::make_pair(a->type, b->type)](
+        a, b, m_intersectionCaches[std::make_pair(a->type, b->type)].get());
 }
 
-void pegasus::geometry::Box::GetAxes(glm::dvec3& a, glm::dvec3& b, glm::dvec3& c) const
+double pegasus::geometry::SimpleShapeIntersectionDetector::CalculatePenetration(SimpleShape const* a, SimpleShape const* b)
 {
-    a = m_aAxis;
-    b = m_bAxis;
-    c = m_cAxis;
+    return m_calculatePenetrationFunctors[std::make_pair(a->type, b->type)](
+        a, b, m_intersectionCaches[std::make_pair(a->type, b->type)].get());
 }
 
-size_t pegasus::geometry::ShapeTypePairHash::operator()(ShapeTypePair const& p) const
+size_t pegasus::geometry::SimpleShapeIntersectionDetector::ShapeTypePairHasher::operator()(ShapeTypePair const& p) const
```

Comment:
Do I understand correctly, that functions `Initialize`, `CalculateIntersection`, `CalculateContactNormal` and `CalculatePenetration` behave the same regardless of what order arguments are given in? In other words, is `Initialize<Sphere, Plane>` equivalent to `Initialize<Plane, Sphere>`?

If so, wouldn't it make more sense to make this hasher order-insensitive? So that, for instance `Plane, Sphere` pair would be hashed the same as `Sphere, Plane`. This could reduce the number of mappings in `SimpleShapeIntersectionDetector`.